sonar-python-plugin/src/main/java/org/sonar/plugins/python/ruff/RuffJsonReportReader.java

```diff
@@ -0,0 +1,111 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.python.ruff;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import javax.annotation.Nullable;
+
+import org.sonarsource.analyzer.commons.internal.json.simple.JSONArray;
+import org.sonarsource.analyzer.commons.internal.json.simple.JSONObject;
+import org.sonarsource.analyzer.commons.internal.json.simple.parser.JSONParser;
+import org.sonarsource.analyzer.commons.internal.json.simple.parser.ParseException;
+
+public class RuffJsonReportReader {
+
+  private final Consumer<Issue> consumer;
+  private final JSONParser jsonParser = new JSONParser();
+
+  public static class Issue {
+    @Nullable
+    String ruleKey;
+    @Nullable
+    String filePath;
+    @Nullable
+    String message;
+    @Nullable
+    Integer startLocationRow;
+    @Nullable
+    Integer startLocationCol;
+    @Nullable
+    Integer endLocationRow;
+    @Nullable
+    Integer endLocationCol;
+  }
+
+  private RuffJsonReportReader(Consumer<Issue> consumer) {
+    this.consumer = consumer;
+  }
+
+  static void read(InputStream in, Consumer<Issue> consumer) throws IOException, ParseException {
+    new RuffJsonReportReader(consumer).read(in);
+  }
+
+  private void read(InputStream in) throws IOException, ParseException {
+    JSONArray files =(JSONArray) jsonParser.parse(new InputStreamReader(in, UTF_8));
+    if (files != null) {
+      ((Stream<JSONObject>) files.stream()).forEach(this::onResult);
+    }
+  }
+
+  private void onResult(JSONObject result) {
+    RuffJsonReportReader.Issue issue = new RuffJsonReportReader.Issue();
+    issue.ruleKey = (String) result.get("code");
+    issue.filePath = (String) result.get("filename");
+    issue.message = (String) result.get("message");
+    JSONObject location = (JSONObject) result.get("location");
+    issue.startLocationCol = toInteger(location.get("column"));
+    issue.startLocationRow = toInteger(location.get("row"));
+    JSONObject endLocation = (JSONObject) result.get("end_location");
+    issue.endLocationCol = correctEndLocationCol(endLocation.get("column"), issue.startLocationCol);
+    issue.endLocationRow = toInteger(endLocation.get("row"));
+    consumer.accept(issue);
+  }
+
+  /*
+    Ruff returns the col number of the last char + 1.
+    In order to properly read the col number we need to return the number of the last char.
+   */
+  private static Integer correctEndLocationCol(Object value, int startLocationCol){
+    Integer endLocationCol = toInteger(value);
+    if(endLocationCol != null){
+      if(endLocationCol > startLocationCol) {
+        return endLocationCol - 1;
+      } else {
+        return startLocationCol;
+      }
+    }
+    return null;
```

Comment:
why can't just return `startLocationCol` if `endLocationCol ` is `null` or  less than `startLocationCol`?

Reply:
If it is null I do not  want to return startLocationCol I want to return null